scripts/pylib/pytest-twister-harness/src/twister_harness/device/device_adapter.py

```diff
@@ -0,0 +1,298 @@
+# Copyright (c) 2023 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: Apache-2.0
+
+from __future__ import annotations
+
+import abc
+import logging
+import os
+import queue
+import re
+import shutil
+import threading
+import time
+from datetime import datetime
+from pathlib import Path
+
+from twister_harness.exceptions import (
+    TwisterHarnessException,
+    TwisterHarnessTimeoutException,
+)
+from twister_harness.twister_harness_config import DeviceConfig
+
+logger = logging.getLogger(__name__)
+
+
+class DeviceAdapter(abc.ABC):
+    """Class defines an interface for all devices."""
+
+    def __init__(self, device_config: DeviceConfig, **kwargs) -> None:
+        """
+        :param device_config: device configuration
+        :param test_name: this information is used only in header log file, it
+            shouldn't impact on adapter code logic
+        """
+        self.device_config: DeviceConfig = device_config
+        self.base_timeout: float = device_config.base_timeout
+        self._device_read_queue: queue.Queue = queue.Queue()
+        self._reader_thread: threading.Thread | None = None
+        self._device_run: threading.Event = threading.Event()
+        self._device_connected: threading.Event = threading.Event()
+        self.command: list[str] = []
+        self._west: str | None = None
+
+        self.handler_log_path: Path = device_config.build_dir / 'handler.log'
+        self._log_files: list[Path] = [self.handler_log_path]
+
+    def __repr__(self) -> str:
+        return f'{self.__class__.__name__}()'
+
+    @property
+    def env(self) -> dict[str, str]:
+        env = os.environ.copy()
+        return env
+
+    def launch(self) -> None:
+        """
+        Start by closing previously running application (no effect if not
+        needed). Then, flash and run test application. Finally, start a reader
+        thread capturing an output from a device.
+        """
+        self.close()
+        self._clear_internal_resources()
+
+        if not self.command:
+            self.generate_command()
+        self._flash_and_run()
+        self._device_run.set()
+        self._start_reader_thread()
+        self.connect()
+
+    def close(self) -> None:
+        """Disconnect, close device and close reader thread."""
+        if not self._device_run.is_set():
+            # device already closed
+            return
+        self.disconnect()
+        self._close_device()
+        self._device_run.clear()
+        self._join_reader_thread()
+
+    def connect(self) -> None:
+        """Connect to device - allow for output gathering."""
+        if not self.is_device_running():
+            msg = 'Cannot connect to not working device'
+            logger.error(msg)
+            raise TwisterHarnessException(msg)
+        if self.is_device_connected():
+            logger.warning('Device already connected')
```

Comment:
Do not log `warning`, `debug` would be enough (or no log). IMO it is not a problem, to connect second time just to be sure, before calling any command from script.
It could be also good to move this check to the beginning of this method.

Reply:
Done